### PR TITLE
Add exception handler #20

### PR DESCRIPTION
**Pull request by svgogin**
Add exception handler #20
1) Exception handler added
2) Logic moved from the controller to the service
3) Custom exceptions added
4) Tests updated

### CODE REVIEW

File: src/main/java/ru/svgogin/service/spark/errordto/CompanyErrorDto.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  @NonNull
+  private final String message;
+
+  public CompanyErrorDto(@NonNull ErrorCode code, @NonNull String message) {
+
+    this.code = code;
+    this.message = message;
+  }
+
+  @NonNull
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/errordto/CompanyErrorDto.java]: this is useless. You can remove it

File: src/main/java/ru/svgogin/service/spark/errordto/CompanyErrorDto.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package ru.svgogin.service.spark.errordto;
+
+import org.springframework.lang.NonNull;
+
+public final class CompanyErrorDto {
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/errordto/CompanyErrorDto.java]:
```suggestion
public final class ErrorDto {
```
This class doesn't depend on company

File: src/main/java/ru/svgogin/service/spark/errordto/CompanyErrorDto.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  @NonNull
+  private final ErrorCode code;
+  @NonNull
+  private final String message;
+
+  public CompanyErrorDto(@NonNull ErrorCode code, @NonNull String message) {
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/errordto/CompanyErrorDto.java]: This annotation doesn't work here
```suggestion
  public CompanyErrorDto(ErrorCode code, String message) {
```

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import ru.svgogin.service.spark.exception.EntityAlreadyExistsException;
+import ru.svgogin.service.spark.exception.NoSuchEntityException;
+
+@ControllerAdvice
+public class RestResponseEntityExceptionHandler extends ResponseEntityExceptionHandler {
+  private static final Logger log = LoggerFactory
+      .getLogger(RestResponseEntityExceptionHandler.class);
+
+  @ExceptionHandler(value
+                        = {EntityAlreadyExistsException.class, NoSuchEntityException.class})
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: nit: it is better to do it this way:
```suggestion
  @ExceptionHandler(value = {
      EntityAlreadyExistsException.class, NoSuchEntityException.class
  })
```

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  protected ResponseEntity<Object> handleConflict(
+      RuntimeException ex, WebRequest request) {
+    String message = ex.getMessage();
+    log.info(message);
+    if (ex instanceof EntityAlreadyExistsException) {
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: you can split exception handlers

File: src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java [hunk 6]
```diff
@@ -198,19 +199,19 @@ void updateShouldReturn404IfNoCompanyForUpdate() throws Exception {
...
     //then
-    Mockito.verify(sparkService).delete(argThat(id -> id.equals(BigInteger.valueOf(1))));
+    Mockito.verify(sparkService).delete(argThat(inn -> inn.equals("07725038124")));
   }
 
   @Test
   void deleteCompanyShouldReturn404ifNotExist() throws Exception {
     // given
-    Mockito.when(sparkService.findByInn(any())).thenReturn(Optional.empty());
+    Mockito.when(sparkService.delete(any())).thenThrow(NoSuchEntityException.class);
```
Comment (alinkov) [inline src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java]: nit: you can use static imports

File: src/test/java/ru/svgogin/service/spark/service/SparkServiceTest.java [hunk 7]
```diff
@@ -111,8 +111,13 @@ void findAllShouldReturnDto() {
   @Test
   @DisplayName("findByInn should call RepositoryDb.findByInn")
   void findByInnShouldCallRepository() {
+    // given
+    String inn = "9705113553";
+    Mockito.when(sparkRepositoryDb.findByInn(inn))
+        .thenReturn(java.util.Optional.of(test7tec)
```
Comment (alinkov) [inline src/test/java/ru/svgogin/service/spark/service/SparkServiceTest.java]:
```suggestion
        .thenReturn(Optional.of(test7tec));
```

File: src/main/java/ru/svgogin/service/spark/service/SparkService.java [hunk 8]
```diff
@@ -58,8 +72,16 @@ public CompanyDto save(CompanyDto companyDto) {
...
-  public void delete(BigInteger id) {
-    sparkRepositoryDb.deleteById(id);
+  public CompanyDto delete(String inn) {
+    var companyOptional = sparkRepositoryDb.findByInn(inn);
+    if (companyOptional.isPresent()) {
+      sparkRepositoryDb.deleteById(companyOptional.get().getId());
+      log.info("Company with inn " + companyOptional.get().getInn() + " was successfully deleted.");
+      return toDto(companyOptional.get());
+    }
+    throw new NoSuchEntityException("Company with inn "
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/service/SparkService.java]: why do you use 3 lines?

File: src/main/java/ru/svgogin/service/spark/service/SparkService.java [hunk 9]
```diff
@@ -58,8 +72,16 @@ public CompanyDto save(CompanyDto companyDto) {
...
+                                    + inn
+                                    + " doesn't exist");
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/service/SparkService.java]:
```suggestion
    var companyDto = sparkRepositoryDb.findByInn(inn)
        .map(this::toDto)
        .orElseThrow(() -> new NoSuchEntityException("Company with inn " + inn + " doesn't exist"));
    sparkRepositoryDb.deleteById(companyDto.getId());
      log.info("Company with inn " + companyDto.getInn() + " was successfully deleted.");
      return companyDto;
    }
```

File: src/main/java/ru/svgogin/service/spark/service/SparkService.java [hunk 10]
```diff
@@ -29,26 +28,41 @@ public SparkService(SparkRepositoryDb sparkRepositoryDb,
...
   }
 
-  public CompanyDto update(CompanyDto companyFromDb, CompanyDto companyDto) {
-    Company companyForUpdate = actualize(companyFromDb, companyDto);
-    return toDto(aggregateTemplate.update(companyForUpdate));
+  public CompanyDto update(String inn, CompanyDto companyDto) {
+    var companyOptional = sparkRepositoryDb.findByInn(inn).map(this::toDto);
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/service/SparkService.java]: use `orElseThrow`

File: src/main/java/ru/svgogin/service/spark/errordto/CompanyErrorDto.java (hunk 1 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/errordto/CompanyErrorDto.java]: Done

File: src/main/java/ru/svgogin/service/spark/errordto/CompanyErrorDto.java (hunk 2 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/errordto/CompanyErrorDto.java]: Done

File: src/main/java/ru/svgogin/service/spark/errordto/CompanyErrorDto.java (hunk 3 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/errordto/CompanyErrorDto.java]: Done

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java (hunk 4 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: Done

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java (hunk 5 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: Done

File: src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java (hunk 6 above)
Comment (svgogin) [reply src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java]: Ok. I've read the infromation about static imports. Is there any simple way in IDE to make all the methods static?

File: src/test/java/ru/svgogin/service/spark/service/SparkServiceTest.java (hunk 7 above)
Comment (svgogin) [reply src/test/java/ru/svgogin/service/spark/service/SparkServiceTest.java]: Done

File: src/main/java/ru/svgogin/service/spark/service/SparkService.java (hunk 8 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/service/SparkService.java]: Done

File: src/main/java/ru/svgogin/service/spark/service/SparkService.java (hunk 10 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/service/SparkService.java]: Done

File: src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java (hunk 6 above)
Comment (alinkov) [reply src/test/java/ru/svgogin/service/spark/web/SparkControllerTest.java]: There is no way to make all your methods static, but you can do it for every case

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+@ControllerAdvice
+public class RestResponseEntityExceptionHandler extends ResponseEntityExceptionHandler {
+  private static final Logger log = LoggerFactory
+      .getLogger(RestResponseEntityExceptionHandler.class);
+
+  @ExceptionHandler(value = {
+      EntityAlreadyExistsException.class
+  })
+
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]:
```suggestion

```

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  protected ResponseEntity<Object> handleConflict(
+      RuntimeException ex, WebRequest request) {
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]:
```suggestion
  protected ResponseEntity<Object> handleConflict(RuntimeException ex, WebRequest request) {
```

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    log.info(message);
+    var bodyOfResponse = new ErrorDto(ErrorDto.ErrorCode.ERROR001, message);
+    return handleExceptionInternal(ex, bodyOfResponse,
+        new HttpHeaders(), HttpStatus.CONFLICT, request);
+  }
+
+  @ExceptionHandler(value = {
+      NoSuchEntityException.class
+  })
+
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]:
```suggestion
  @ExceptionHandler(value = {NoSuchEntityException.class})
```

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+  protected ResponseEntity<Object> handleNotFound(
+      RuntimeException ex, WebRequest request) {
+    String message = ex.getMessage();
+    log.info(message);
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: may be use `log.warn`?

File: src/main/java/ru/svgogin/service/spark/service/SparkService.java [hunk 15]
```diff
@@ -29,26 +28,37 @@ public SparkService(SparkRepositoryDb sparkRepositoryDb,
...
 
-  public Optional<CompanyDto> findByInn(String inn) {
-    return sparkRepositoryDb.findByInn(inn).map(this::toDto);
+  public CompanyDto findByInn(String inn) {
+    var companyDto = sparkRepositoryDb.findByInn(inn)
+        .map(this::toDto)
+        .orElseThrow(() -> new NoSuchEntityException("Company with inn " + inn + " doesn't exist"));
+    log.info("Company with inn "
+             + companyDto.getInn()
+             + " was successfully returned.");
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/service/SparkService.java]: Why do you use three lines for this? This makes your code more complicated
```suggestion
    log.info("Company with inn " + companyDto.getInn() + " was successfully returned.");
```

File: src/test/java/ru/svgogin/service/spark/service/SparkServiceTest.java [hunk 16]
```diff
@@ -111,8 +112,13 @@ void findAllShouldReturnDto() {
   @Test
   @DisplayName("findByInn should call RepositoryDb.findByInn")
   void findByInnShouldCallRepository() {
+    // given
+    String inn = "9705113553";
+    Mockito.when(sparkRepositoryDb.findByInn(inn))
+        .thenReturn(Optional.of(test7tec)
+        );
```
Comment (alinkov) [inline src/test/java/ru/svgogin/service/spark/service/SparkServiceTest.java]:
```suggestion
        .thenReturn(Optional.of(test7tec));
```

File: src/main/java/ru/svgogin/service/spark/service/SparkService.java (hunk 9 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/service/SparkService.java]: Done

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java (hunk 11 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: Done

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java (hunk 12 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: Done

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java (hunk 13 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: Done

File: src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java (hunk 14 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/handler/RestResponseEntityExceptionHandler.java]: Done

File: src/main/java/ru/svgogin/service/spark/service/SparkService.java (hunk 15 above)
Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/service/SparkService.java]: Sometimes I get a CheckStyle warning: "A line length is more than 100 signs"

Comment (svgogin) [reply src/main/java/ru/svgogin/service/spark/service/SparkService.java]: Done

File: src/test/java/ru/svgogin/service/spark/service/SparkServiceTest.java (hunk 16 above)
Comment (svgogin) [reply src/test/java/ru/svgogin/service/spark/service/SparkServiceTest.java]: Done

File: src/main/java/ru/svgogin/service/spark/service/SparkService.java [hunk 17]
```diff
@@ -58,8 +66,13 @@ public CompanyDto save(CompanyDto companyDto) {
...
   }
 
-  public void delete(BigInteger id) {
-    sparkRepositoryDb.deleteById(id);
+  public CompanyDto delete(String inn) {
+    var companyDto = sparkRepositoryDb.findByInn(inn)
+        .map(this::toDto)
+        .orElseThrow(() -> new NoSuchEntityException("Company with inn " + inn + " doesn't exist"));
+    sparkRepositoryDb.deleteById(companyDto.getId());
+    log.info("Company with inn " + companyDto.getInn() + " was successfully deleted.");
```
Comment (alinkov) [inline src/main/java/ru/svgogin/service/spark/service/SparkService.java:74]: Logger support this formatting:
```java
log.info("Company with inn {} was successfully deleted.", companyDto.getInn());
```

**Review by alinkov — APPROVED** [review]

**Pull request merged**